File: sdk-py/src/capsule/chain.py
"""Event chain build + verify. Mirrors sdk/src/chain.js."""

from __future__ import annotations

import json
from typing import TypedDict

from .canonical import bytes_to_hex, concat_bytes, hex_to_bytes, jcs, sha256

GENESIS_PREV_BYTES: bytes = b"\x00" * 32
GENESIS_PREV_HEX: str = "0" * 64


class ChainError(TypedDict):
    seq: int
    message: str


class ChainResult(TypedDict):
    ok: bool
    errors: list[ChainError]


def hash_event(event: dict) -> bytes:
    """Compute event hash. event must NOT include 'hash'; prev_hash must be 64-hex.
    Returns 32 bytes."""
    if "hash" in event:
        raise ValueError("hash_event: event must not include 'hash'")
    prev_hex = event.get("prev_hash")
    if not isinstance(prev_hex, str) or len(prev_hex) != 64:
        raise ValueError("hash_event: prev_hash must be 64-hex")
    prev_raw = hex_to_bytes(prev_hex)
    canonical = jcs(event)
    return sha256(concat_bytes(prev_raw, canonical))


def build_chain_events(bare_events: list[dict]) -> list[dict]:
    """Walk a list of bare events and assign prev_hash + hash + seq + event_id."""
    out: list[dict] = []
    prev = GENESIS_PREV_BYTES
    for i, bare in enumerate(bare_events):
        seq = i + 1
        event_id = bare.get("event_id") or f"evt_{seq:03d}"
        e: dict = {
            "seq": seq,
            "event_id": event_id,
            **{k: v for k, v in bare.items() if k != "event_id"},
            "prev_hash": bytes_to_hex(prev),
        }
        if "payload" not in e or e["payload"] is None:
            e["payload"] = {}
        if not isinstance(e.get("untrusted_payload_fields"), list):
            cands: list[str] = []
            payload = e["payload"]
            if isinstance(payload, dict):
                if isinstance(payload.get("summary"), str):
                    cands.append("payload.summary")
                if isinstance(payload.get("statement"), str):
                    cands.append("payload.statement")
            e["untrusted_payload_fields"] = cands
        h = hash_event(e)
        e["hash"] = bytes_to_hex(h)
        out.append(e)
        prev = h
    return out


def events_to_jsonl(events: list[dict]) -> bytes:
    """Serialize built events into JSONL bytes."""
    lines = [json.dumps(e, separators=(",", ":"), ensure_ascii=False) for e in events]
    return ("\n".join(lines) + "\n").encode("utf-8")


def events_from_jsonl(data: bytes) -> list[dict]:
    """Parse JSONL bytes into events."""
    text = data.decode("utf-8")
    out = []
    for i, line in enumerate(text.split("\n")):
        if not line:
            continue
        try:
            out.append(json.loads(line))
        except json.JSONDecodeError as ex:
            raise ValueError(f"chain line {i + 1}: invalid JSON: {ex.msg}") from ex
    return out


def verify_chain(events: list[dict]) -> ChainResult:
    """Verify a chain. Returns ChainResult with ok and collected errors."""
    errors: list[ChainError] = []
    prev = GENESIS_PREV_BYTES
    for i, e in enumerate(events):
        seq = e.get("seq", i + 1)
        if e.get("seq") != i + 1:
            errors.append({"seq": seq, "message": f"seq {e.get('seq')} expected {i + 1}"})
        if not isinstance(e.get("prev_hash"), str) or len(e["prev_hash"]) != 64:
            errors.append({"seq": seq, "message": "prev_hash missing or wrong length"})
            continue
        expected_prev = bytes_to_hex(prev)
        if e["prev_hash"] != expected_prev:
            errors.append(
                {
                    "seq": seq,
                    "message": f"prev_hash mismatch: got {e['prev_hash']}, expected {expected_prev}",
                }
            )
        if not isinstance(e.get("hash"), str) or len(e["hash"]) != 64:
            errors.append({"seq": seq, "message": "hash missing or wrong length"})
            continue
        rest = {k: v for k, v in e.items() if k != "hash"}
        try:
            recomputed = bytes_to_hex(hash_event(rest))
        except ValueError as ex:
            errors.append({"seq": seq, "message": f"recompute failed: {ex}"})
            continue
        if recomputed != e["hash"]:
            errors.append(
                {
                    "seq": seq,
                    "message": f"hash mismatch: stored {e['hash']}, recomputed {recomputed}",
                }
            )
        prev = hex_to_bytes(e["hash"])
    return {"ok": len(errors) == 0, "errors": errors}


def first_and_entry_hash(events: list[dict]) -> tuple[str, str]:
    """Return (first_event_hash, entry_hash) for the chain."""
    if not events:
        raise ValueError("chain is empty")
    return events[0]["hash"], events[-1]["hash"]

File: sdk-js/src/chain.js
// Event chain: hashing with raw bytes, no hex strings as inputs.

import { bytesToHex, concatBytes, hexToBytes, jcs, sha256 } from "./canonical.js";

const GENESIS_PREV = Buffer.alloc(32, 0);

/**
 * Compute event hash. event must NOT include "hash"; "prev_hash" must be hex.
 * Returns 32-byte Buffer.
 */
export function hashEvent(event) {
  if (event.hash !== undefined) throw new Error("hashEvent: event must not include 'hash'");
  if (typeof event.prev_hash !== "string" || event.prev_hash.length !== 64) {
    throw new Error("hashEvent: prev_hash must be 64-hex");
  }
  const prevRaw = hexToBytes(event.prev_hash);
  const canon = jcs(event);
  return sha256(concatBytes(prevRaw, canon));
}

/**
 * Walk a list of (mostly) bare events and assign prev_hash + hash + seq + event_id.
 * Events should already have actor/kind/action/target/timestamp/payload set.
 */
export function buildChainEvents(bareEvents) {
  let prev = GENESIS_PREV;
  const out = [];
  bareEvents.forEach((bare, i) => {
    const seq = i + 1;
    const event_id = bare.event_id ?? `evt_${String(seq).padStart(3, "0")}`;
    const e = {
      seq,
      event_id,
      ...bare,
      prev_hash: bytesToHex(prev),
    };
    if (e.payload === undefined) e.payload = {};
    if (!Array.isArray(e.untrusted_payload_fields)) {
      // default: mark common LLM-narrative fields untrusted if present
      const candidates = [];
      if (typeof e.payload?.summary === "string") candidates.push("payload.summary");
      if (typeof e.payload?.statement === "string") candidates.push("payload.statement");
      e.untrusted_payload_fields = candidates;
    }
    const h = hashEvent(e);
    e.hash = bytesToHex(h);
    out.push(e);
    prev = h;
  });
  return out;
}

/** Serialize built events into JSONL bytes. */
export function eventsToJsonl(events) {
  const lines = events.map((e) => JSON.stringify(e));
  return Buffer.from(lines.join("\n") + "\n", "utf8");
}

/** Parse JSONL bytes into events. */
export function eventsFromJsonl(bytes) {
  const text = Buffer.from(bytes).toString("utf8");
  const lines = text.split("\n").filter((l) => l.length > 0);
  return lines.map((line, i) => {
    try {
      return JSON.parse(line);
    } catch (err) {
      throw new Error(`chain line ${i + 1}: invalid JSON: ${err.message}`);
    }
  });
}

/** Verify a chain. Returns { ok, errors: [{ seq, message }] }. */
export function verifyChain(events) {
  const errors = [];
  let prev = GENESIS_PREV;
  events.forEach((e, i) => {
    const seq = e.seq ?? i + 1;
    if (e.seq !== i + 1) {
      errors.push({ seq, message: `seq ${e.seq} expected ${i + 1}` });
    }
    if (typeof e.prev_hash !== "string" || e.prev_hash.length !== 64) {
      errors.push({ seq, message: "prev_hash missing or wrong length" });
      return;
    }
    const expectedPrev = bytesToHex(prev);
    if (e.prev_hash !== expectedPrev) {
      errors.push({
        seq,
        message: `prev_hash mismatch: got ${e.prev_hash}, expected ${expectedPrev}`,
      });
    }
    if (typeof e.hash !== "string" || e.hash.length !== 64) {
      errors.push({ seq, message: "hash missing or wrong length" });
      return;
    }
    const { hash, ...rest } = e;
    let recomputedHex;
    try {
      const recomputed = hashEvent(rest);
      recomputedHex = bytesToHex(recomputed);
    } catch (err) {
      errors.push({ seq, message: `recompute failed: ${err.message}` });
      return;
    }
    if (recomputedHex !== hash) {
      errors.push({
        seq,
        message: `hash mismatch: stored ${hash}, recomputed ${recomputedHex}`,
      });
    }
    prev = hexToBytes(hash);
  });
  return { ok: errors.length === 0, errors };
}

export function firstAndEntryHash(events) {
  if (events.length === 0) throw new Error("chain is empty");
  return {
    firstEventHash: events[0].hash,
    entryHash: events[events.length - 1].hash,
  };
}
